tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/ViolationTest.java

```diff
@@ -0,0 +1,77 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.AbstractActionSequenceElement;
+
+public class ViolationTest extends AnalysisFeatureTest {
+	/**
+     * Tests, whether the analysis correctly identifies violations for the travel planner model
+     * <p>
+     * Fails if the analysis does not propagate the correct characteristics for each ActionSequence
+     */
+    @DisplayName("Find violations for the example model")
+    @MethodSource("violationTestProvider")
+    @ParameterizedTest
+    public void violationTest(StandalonePCMDataFlowConfidentialtyAnalysis analysis, Predicate<AbstractActionSequenceElement<?>> contraint, boolean noViolations) {
+        var sequences = analysis.findAllSequences();
+        var propagationResult = analysis.evaluateDataFlows(sequences);
+
+        var result = analysis.queryDataFlow(propagationResult.get(0), contraint);
+        assertEquals(noViolations, result.isEmpty());
+    }
+    
+    private Stream<Arguments> violationTestProvider() {
+    	Predicate<AbstractActionSequenceElement<?>> travelPlannerContraint = node -> travelPlannerCondition(node);
+    	Predicate<AbstractActionSequenceElement<?>> internationalOnlineShopContraint = node -> internationalOnlineShopCondition(node);
+    	return Stream.of(
+    			Arguments.of(travelPlannerAnalysis, travelPlannerContraint, true),
+    			Arguments.of(internationalOnlineShopAnalysis, internationalOnlineShopContraint, false)
+    			);
+    }
+    
+    /**
+     * Indicates whether an element in an action sequence violates the constraint of the travel planner model
+     * @param node Element of the action sequence
+     * @return Returns true, if the constraint is violated. Otherwise, the method returns false.
+     */
+    private boolean travelPlannerCondition(AbstractActionSequenceElement<?> node) {
+    	var assignedRoles = node.getAllNodeVariables().stream()
+    			.filter(cv -> cv.characteristicType().getName().equals("AssignedRoles"))
+    			.map(cv -> cv.characteristicLiteral())
+    			.collect(Collectors.toList());
+    	var grantedRoles = node.getAllDataFlowVariables().stream()
+    			.flatMap(df -> df.getAllCharacteristics().stream())
+    			.filter(cv -> cv.characteristicType().getName().equals("GrantedRoles"))
+    			.map(cv -> cv.characteristicLiteral())
+    			.collect(Collectors.toList());
+    	return assignedRoles.stream().noneMatch(ar -> grantedRoles.contains(ar));
+    }
+    
+    /**
+     * Indicates whether an element in an action sequence violates the constraint of the international online shop model
+     * @param node Element of the action sequence
+     * @return Returns true, if the constraint is violated. Otherwise, the method returns false.
+     */
+    private boolean internationalOnlineShopCondition(AbstractActionSequenceElement<?> node) {
```

Comment:
Props that you were able to write down that constraint without help. Looking at the retrieval of the information saved in `serverLocation` and `dataSensitivity`: I would argue that such filtering operations (retrieving literals by name on a node) will be so common, we should it include it as public methods directly in the node.